runtime/lua/vim/lsp.lua

```diff
@@ -2040,7 +2040,7 @@ end
 ---
 ---@return table[] tags A list of matching tags
 function lsp.tagfunc(pattern, flags)
-  return vim.lsp.tagfunc(pattern, flags)
+  return vim.lsp._tagfunc(pattern, flags)
```

Comment:
The purpose of this function was originally to make calling tagfunc easier (without needing `require`), but now that that isn’t necessary, this function is just a useless wrapper. Can we keep the tagfunc module named as-is and remove this wrapper function instead?

Reply:
> this function is just a useless wrapper

That's true, I also felt the same but it was bothering me that `vim.lsp.tagfunc` is a module and function (returned by the module). The problem is `gen_vimdoc`: it does not recognize the docs moved to `vim.lsp.tagfunc` --- but I fully agree with you. Let me find a solution to make docgen happy.

Reply:
If we move docs to the `vim.lsp.tagfunc` module and remove the wrapper, a new *section* in vimdoc will be added:

```
Lua module: vim.lsp.tagfunc                                      *lsp-tagfunc*

tagfunc                                              *vim.lsp.tagfunc.tagfunc*
    Provides an interface between the built-in client and 'tagfunc'.
```

But I'm not sure we want this. (BTW: it should be overridden to `*vim.lsp.tagfunc*` ...)

Reply:
This works, though it's a bit of a hack...

```diff
diff --git a/runtime/lua/vim/lsp.lua b/runtime/lua/vim/lsp.lua
index 5fa5a1db2..29f6eb8c4 100644
--- a/runtime/lua/vim/lsp.lua
+++ b/runtime/lua/vim/lsp.lua
@@ -2040,9 +2040,12 @@ end
 ---
 ---@return table[] tags A list of matching tags
 function lsp.tagfunc(pattern, flags)
-  return vim.lsp.tagfunc(pattern, flags)
+  error('Unreachable')
 end
 
+---@nodoc
+lsp.tagfunc = nil
+
 ---Checks whether a client is stopped.
 ---
 ---@param client_id (integer)
```

Reply:
Yeah, nice idea; looks hacky :D. The useless --- but useful only for the doc --- wrapper would be better though..

Reply:
(Just leaving my 2 cents here) I think that dirty hacks are neat when:
- The code documents why the hack is needed and why/how it works.
- There's `git` history that provides context information around the problem (and hack).
- (Optional but still lovely) There's a `TODO` comment/GitHub issue/something documenting that this is a hack and that there's future hope for a cleaner non-hacky solution that someone will be brave enough to tackle.

In this particular situation, I believe the hack is worth it but I insist in the conditions above ;)

Reply:
We have a similar issue with `vim.treesitter.foldexpr()` where `foldexpr` is returned by `treesitter._fold`.

This issue is really disrupting my workflow, so I'm just going to merge this as is. We can accept a follow-up PR that either fixes `gen_vimdoc.py` or adds the above "hack", but for now a redundant wrapper to a private module isn't a huge problem.